proto/provenance/reward/v1/reward.proto

```diff
@@ -0,0 +1,188 @@
+  // Used internally to calculate and track the current claim period's ending time.
+  google.protobuf.Timestamp claim_period_end_time = 14 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
+  // Time the RewardProgram switched to FINISHED state. Initially set as empty.
+  google.protobuf.Timestamp actual_program_end_time = 15 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
+  // Number of claim periods this program will run for.
+  uint64 claim_periods = 16;
+  // Current claim period of the RewardProgram. Uses 1-based indexing.
+  uint64 current_claim_period = 17;
+  // maximum number of claim periods a reward program can rollover.
+  uint64 max_rollover_claim_periods = 18;
+  // Current state of the RewardProgram.
+  State state = 19;
+  // Grace period after a RewardProgram FINISHED. It is the number of seconds until a RewardProgram enters the EXPIRED
+  // state.
+  uint64 expiration_offset = 20;
+  // Actions that count towards the reward.
+  repeated QualifyingAction qualifying_actions = 21 [(gogoproto.nullable) = false];
+}
+
+// ClaimPeriodRewardDistribution, this is updated at the end of every claim period.
+message ClaimPeriodRewardDistribution {
+  option (gogoproto.equal)            = true;
+  option (gogoproto.goproto_stringer) = true;
+
+  // The claim period id.
+  uint64 claim_period_id = 1;
+  // The id of the reward program that this reward belongs to.
+  uint64 reward_program_id = 2;
+  // The sum of all the granted rewards for this claim period.
+  cosmos.base.v1beta1.Coin total_rewards_pool_for_claim_period = 3 [(gogoproto.nullable) = false];
+  // The final allocated rewards for this claim period.
+  cosmos.base.v1beta1.Coin rewards_pool = 4 [(gogoproto.nullable) = false];
+  // The total number of granted shares for this claim period.
+  int64 total_shares = 5;
+  // A flag representing if the claim period for this reward has ended.
+  bool claim_period_ended = 6;
+}
+
+// RewardAccountState contains state at the claim period level for a specific address.
+message RewardAccountState {
+  option (gogoproto.equal)            = true;
+  option (gogoproto.goproto_stringer) = true;
+  // ClaimStatus is the state a claim is in
+  enum ClaimStatus {
+    // undefined state
+    CLAIM_STATUS_UNSPECIFIED = 0;
+    // unclaimable status
+    CLAIM_STATUS_UNCLAIMABLE = 1;
+    // unclaimable claimable
+    CLAIM_STATUS_CLAIMABLE = 2;
+    // unclaimable claimed
+    CLAIM_STATUS_CLAIMED = 3;
+    // unclaimable expired
+    CLAIM_STATUS_EXPIRED = 4;
+  }
+  // The id of the reward program that this share belongs to.
+  uint64 reward_program_id = 1;
+  // The id of the claim period that the share belongs to.
+  uint64 claim_period_id = 2;
+  // Owner of the reward account state.
+  string address = 3;
+  // The number of actions done by this account, mapped by action type.
+  map<string, uint64> action_counter = 4;
```

Comment:
I'm pretty sure we cannot use maps in state stuff due to non-deterministic ordering.

Reply:
Yeah.
https://developers.google.com/protocol-buffers/docs/proto3#maps
> Wire format ordering and map iteration ordering of map values are undefined, so you cannot rely on your map items being in a particular order.
So this will need to be changed to a `repeated` new message with a `string` and `int64`.